### ISSUE TITLE
rpc backend doesn't seem to report message status correctly.

### ISSUE THREAD

**Issue by mke21** [celery/celery#4084]
This is version 4.0.2
If I use rabbitmq, the rpc result backend, and a custom queue the message's status never seems to change and stays 'PENDING', even if the logging in the worker reports that it has been successfully executed until I do a get() of some sorts when the status changes to 'SUCCESS'. When I change the backend to amqp this the system works as expected, giving SUCCESS before doing the get(). Also the redis backend doesn't give this problem, so it seems to be rpc specific.

Note that not setting a custom queue, so using the default also works as expected just like the other backends!

I've got in my tasks.py:
```python
# tasks.py
from celery import Celery

app = Celery('tasks', 
broker='amqp://',
backend='rpc://' # here I use the rpc backend
)
@app.task
def test_task(s):
    return s
```
I start this with `celery -A tasks worker --loglevel=info -Q myqueue

On the other side I do:
```python
>>> import tasks
>>> result = tasks.test_task.apply_async(queue='myqueue', args=['blaat',]) # set custom queue

>>> result.ready() # worker already executed, so should give True
False
>>> result.status # with RPC backend it returned False while it should have been True in previous statement
'PENDING'
>>> print(result.get())
'This is a string'
>>> result.ready()
True
>>> result.status
'SUCCESS'
```

If I start this qithout the -Q option: 'celery -A tasks worker --loglevel=info' I get:
```python
>>> import tasks
>>> result = tasks.test_task.apply_async(args=['blaat',]) # set no custom queue

>>> result.ready() # worker already executed, so should give True
True
>>> result.status # previous seemed to work
'SUCCESS'
>>> print(result.get()) # SUCCESS was also reported
'This is a string'
>>> result.ready()
True
>>> result.status
'SUCCESS'
```

**Comment by jpescalona** [comment]
I've been able to reproduce this issue with:
* Celery 4.0.2
* backend_results = 'rpc://'
but only it start to fail when I execute a group task before individual tasks. Anyway, GroupResult does not return ready even if all subtasks are ready. When I call GroupResult.get it marks GroupResult as ready.
```
In [1]: from myapp.tasks import invalidate_cache
In [2]: r = invalidate_cache.apply_async(queue='management')

In [3]: r.state
Out[3]: u'STARTED'

In [4]: r.state
Out[4]: u'STARTED'

In [5]: r.state
Out[5]: u'STARTED'

In [6]: r.ready()
Out[6]: True

In [7]: r.ready()
Out[7]: True

In [8]: r.state
Out[8]: u'SUCCESS'

In [9]: from myapp.tasks import invalidate_cache

In [10]: from celery import group

In [11]: g = group([invalidate_cache.s().set(queue='management')]).apply_async()

In [12]: g.ready()
Out[12]: False

In [13]: g.ready()
Out[13]: False

In [14]: g.ready()
Out[14]: False

In [15]: g.ready()
Out[15]: False

In [16]: g.ready()
Out[16]: False

In [17]: g.get()
Out[17]: [None]

In [18]: g.ready()
Out[18]: True

In [19]: r = invalidate_cache.apply_async(queue='management')

In [20]: r.state
Out[20]: u'PENDING'

In [21]: r.state
Out[21]: u'PENDING'

In [22]: r.state
Out[22]: u'PENDING'

In [23]: r.get()

In [24]: r.state
Out[24]: u'SUCCESS'
```

**Comment by jpescalona** [comment]
After digging a while, i've discovered that this problem is reproduced when using rpc as results_backend instead of amqp. Maybe rpc results backend is broken?

**Comment by thedrow** [comment]
This sounds like a bug. Thanks for the report.
Did any of you try doing the same with master?

**Comment by jpescalona** [comment]
I've tried against 4.0.2, 4.1.0 and master, and for all of them, rpc backend do not update the tasks status correctly.

**Comment by donkopotamus** [comment]
I have also observed something similar to this issue, where despite a worker being configure to send a STARTED message (via `task_send_started`), and having confirmed (by inserting a breakpoint) that it *is* being sent, the rpc backend does not appear to process it with the `AsyncResult.state` moving only between PENDING to SUCCESS.

Currently I have traced the issue to this piece of code in `celery.backends.async:BaseResultConsumer`:

```
    def on_state_change(self, meta, message):
        if self.on_message:
            self.on_message(meta)
        if meta['status'] in states.READY_STATES:    ### <===== HERE
            task_id = meta['task_id']
            try:
                result = self._get_pending_result(task_id)
            except KeyError:
                # send to buffer in case we received this result
                # before it was added to _pending_results.
                self._pending_messages.put(task_id, meta)
            else:
                result._maybe_set_cache(meta)
                buckets = self.buckets
                try:
                    # remove bucket for this result, since it's fulfilled
                    bucket = buckets.pop(result)
                except KeyError:
                    pass
                else:
                    # send to waiter via bucket
                    bucket.append(result)
        sleep(0)
```

I have confirmed that the `STARTED` message is being processed by this `on_message` but clearly as it is not in `states.READY_STATES` it will never be acted on in any way.  Indeed other custom messages will also not be processed.

I'm c

**Comment by auvipy** [comment]
could you also check latest master and report?

**Comment by xirdneh** [comment]
Somebody reported this as an issue on IRC.
I can confirm the state and ready values are not correct until you do `result.get()` when using RPC as result backend.
This is using the latest master

**Comment by xirdneh** [comment]
Another update while I was testing this. Apparently it is working for single tasks but the same problem arises when using groups.
```python
>>> res = group(add.s(3,4), add.s(5,7)).delay()
>>> for r in res.results:
...     print(r.state)
PENDING
PENDING
```
This might have to be a new issue...

**Comment by codeSamuraii** [comment]
I'm having a similar issue.

I have a group of task executing correctly when using the AMQP backend. When switching to RPC, the group task never finishes and blocks on the `group_result.get()` statement.

**Comment by pyDev20** [comment]
I encountered an issue that might be related, in version 4.3.0.

I have a simple celery application with two tasks, a_func() and b_func(). After starting the celery worker, I am calling a_func.apply_async(), and a_func, when running on worker is calling b_func.apply_async().

When using 'amqp://' as a backend everything is working well. However, when using 'rpc://' as a backend, I am having problems. I am trying to get the state and the return value of the tasks. For the a_func() task, there is no problem. However for b_func() I am getting state = 'PENDING' forever, and get() is stuck forever.

I am using celery version 4.3.0.

Worker cmd:
`
celery -A celery_test worker --loglevel=inf
`
celery application:
`
app = Celery('my_app',
             backend='rpc://',
             broker='pyamqp://guest@localhost/celery',
             include=['tasks'])
`
a_func and b_func tasks:
`
@task
def a_func():
    print "A"
    b_func.apply_async()
    return "A"

@task
def b_func():
    print "B"
    return "B"
`

**Comment by amard33p** [comment]
+1. This is probably due to this issue : https://github.com/celery/celery/issues/4830 which has a fix in the works.  
Changing the backend to redis (since amqp will be deprecated) solves the issue.

**Comment by shaloba** [comment]
Just encountered with this issue and i noticed the issue is still opened. any news ? 
tnx

**Comment by thedrow** [comment]
It may be a limitation of the RPC backend since it doesn't store state.
We might be able to work around this by storing the task state elsewhere.

I think this issue might only be fixed after our NextGen architecture refactoring.

**Comment by thangdc94** [comment]
I faced the same issue today on celery version 5.1.2

Everything works fine on my windows PC but when I deployed the code on Linux server `GroupResult.get()` stuck forever.

I found out that use celery version 5.0.0 can solve the issue on my Linux server.

Hope this information can help you guys debug the issue.

**Comment by thangdc94** [comment]
> then it might be a regression, can you try 5.2.0rc1 and report back

I tried with celery version 5.2.0rc1 and still faced the same problem.
I installed many older versions and found out that the issue begins at 5.0.3. 
I also tested with version 4.4.7 and got no issue.

**Comment by Venkatesh283** [comment]
I'm facing the same issue with celery versions - 4.4.5, 4.4.7, 5.1.2, 5.2.0cr1, 5.2.7
Worker is configured with Concurrency = 10 / prefork pool. 
Broker - AMQP
Backend - rpc

When a single task added at a time, it works fine. AyncResult.state is updated to SUCCESS.
But in case of multiple tasks added in parallel (5 or more at a time) the workers are successfully completing the tasks.  AyncResult.state is in "PENDING" state even after completion of task. And also the AyncResult.get() stucks forever.

Any fixes or updates?

Thanks in advance.

**Comment by Mortaza-Seydi** [comment]
Hi, 
@Venkatesh283 i have the same problem, `.get()` stucks forever and `.state` is `Pending` but worker logs shows the task is succeeded with return value.
celery = 5.2.7
python = 3.8.10
CELERY_BROKER = 'amqp://guest@localhost//'
CELERY_BACKEND = 'rpc://'

**Comment by extratrees** [comment]
Same problem here. 
Using celery==5.2.7, python==3.8.14 
CELERY_BROKER = 'amqp://guest@localhost//'
CELERY_BACKEND = 'rpc://'.

Will try to switch to Redis, but this adds a layer of complexity

**Comment by Sushanti99** [comment]
Hey. Been meaning to switch to rpc backend for more support for monitoring purposes like Flower. Is it still not safe to upgrade to 5.x and use rpc backend? (Currently using amqp backend but that's deprecated in 5.x)

**Comment by Fagtoy** [comment]
Same problem
celery==5.2.7 and 5.3.0rc1, python==3.10
CELERY_BROKER = 'amqp://guest@localhost//'
CELERY_BACKEND = 'rpc://'

tasks statuses are always pending and `.get()` takes forever

**Comment by buildxyz-git** [comment]
Same issue here using:
```
celery==5.2.7 and python==3.10
CELERY_BROKER = 'amqp://guest@localhost//'
CELERY_BACKEND = 'rpc://'
```
Seems this issue has been lurking for years, I went down the RabbitMQ hole as it seemed to be the recommended celery backend at the time but now it seems the solution is to switch over to another backend like Redis?

**Comment by rms-sth** [comment]
The issue still exists in Celery 5.4.0 with Python 3.11.8

**Comment by Thesnowyegret** [comment]
I believe I came across a different issue that is caused by the same bug that is causing this issue, I also believe that my suggested solution is a fix for both. 

the issue I faced happens when I use `self.update_state` (or `celery.app.task.Task.update_state`) inside a bind=True task, to update my state with a progress tracker.

env :
```
Celery[librabbitmq]==5.4.0 and python == 3.9.13
CELERY_BROKER = 'amqp://'
CELERY_BACKEND = 'rpc://'
```
tasks.py
```
app = Celery()


def _state_progress(self: celery.app.task.Task, current, total, desc=""):
    direct = self.request.called_directly
    if total < 0:
        raise ValueError("total is less than 0")
    if current > total:
        raise ValueError("current is bigger than total")
    if not direct and total >= 0:
        self.update_state(state='PROGRESS', meta={'current': current, 'total': total, 'description': desc})


@app.task(bind=True)
def my_task(self, step_sleep=1, total_steps=5):
    for i in range(total_steps):
        _state_progress(self, i, total_steps, f"stage {i}!")
        time.sleep(step_sleep)
    _state_progress(self, total_steps, total_steps, f"stage {total_steps}!")
    time.sleep(step_sleep)
    return "DONE!"
```
test.py
```
import random
import time

import celery.exceptions
import celery.result

import tasks


def test_block():
    my_id = random.randint(0, 1000)
    print(f"{my_id}: starting blocking!")
    r = tasks.my_task.delay()
    r: celery.result.AsyncResult

    print(f"{my_id}: {r.get()}")


def test_progress():
    my_id = random.randint(0, 1000)
    print(f"{my_id}: starting tracked!")
    r = tasks.my_task.delay()
    r: celery.result.AsyncResult

    current_state = ""
    current_step = -1

    while not r.ready():
        meta = r.info  # this is important for timing reasons
        state = r.state
        if state != current_state:
            current_state = state
            print(f"{my_id}: {current_state}")

        if state == "PROGRESS" and isinstance(meta, dict):
            current_total = meta.get("total", -1)
            step = meta.get("current", -1)
            if step != current_step:
                current_step = step
                print(f"{my_id}: {current_state} {current_step}/{current_total}")
        else:
            try:
                res = r.get(0.0)
                print(f"{my_id}: {res}")
                return
            except celery.exceptions.TimeoutError:
                pass

        time.sleep(0.2)

    print(f"{my_id}: {r.get()}")


if __name__ == '__main__':
    random.seed(10)
    test_progress()
    test_block()
    test_progress()
```
output:
```
585: starting tracked!
585: PROGRESS
585: PROGRESS 0/5
585: PROGRESS 1/5
585: PROGRESS 2/5
585: PROGRESS 3/5
585: PROGRESS 4/5
585: PROGRESS 5/5
585: DONE!
33: starting blocking!
33: DONE!
439: starting tracked!
439: PENDING
439: DONE!
```

if on the publisher side a blocking `result.get()` is used, the` result.ready(), result.info, result.state` stop working!

this is caused by the `on_result_fulfilled()` function in celery.backends.rpc.py (lines 187-190). when `result.get()` is used it creates a consumer that is supposed to be stopped when the `on_result_fulfilled()` is called, but because the method overrides the original implementation with a pass, this consumer is never stopped, and the `AsyncResult._get_task_meta()` stops working.

the fix is simple, delete the `on_result_fulfilled()` function in celery.backends.rpc.py, so that the rpc uses the parents implementation.

output after deleting :
```
585: starting tracked!
585: PROGRESS
585: PROGRESS 0/5
585: PROGRESS 1/5
585: PROGRESS 2/5
585: PROGRESS 3/5
585: PROGRESS 4/5
585: PROGRESS 5/5
585: DONE!
33: starting blocking!
33: DONE!
439: starting tracked!
439: PROGRESS
439: PROGRESS 0/5
439: PROGRESS 1/5
439: PROGRESS 2/5
439: PROGRESS 3/5
439: PROGRESS 4/5
439: PROGRESS 5/5
439: DONE!
```

I have also tested the problem described in this issue and this fix works for it as well.
